prettyprinter-configurable/src/Text/PrettyBy/Fixity.hs

```diff
@@ -29,112 +29,122 @@ import Lens.Micro
+{-| Takes a monadic pretty-printer and turns it into one that receives a @config@ explicitly.
+Useful for defining instances of 'PrettyBy' monadically when writing precedence-aware
+pretty-printing code (and since all functions below are monadic, it's currenty the only option).
+-}
 inContextM :: (a -> InContextM config (Doc ann)) -> config -> a -> Doc ann
 inContextM pM config = runInContextM config . pM
 
 -- | A string written in the 'InContextM' monad gets enclosed with 'unitDocM' automatically.
 instance (HasRenderContext config, doc ~ Doc ann) => IsString (InContextM config doc) where
-    fromString = unitDocM . fromString
+  fromString = unitDocM . fromString
 
 -- TODO: when writing a precedence-aware pretty-printer we basically always want to specify a
 -- fixity in each clause. Would be nice to enforce that in types.
--- | Enclose a 'Doc' in parentheses if required or leave it as is. The need for enclosing is
--- determined from an outer 'RenderContext' (stored in the environment of the monad) and the inner
--- fixity provided as an argument.
-encloseM :: MonadPrettyContext config env m => Fixity -> Doc ann -> m (Doc ann)
+
+{-| Enclose a 'Doc' in parentheses if required or leave it as is. The need for enclosing is
+determined from an outer 'RenderContext' (stored in the environment of the monad) and the inner
+fixity provided as an argument.
+-}
+encloseM :: (MonadPrettyContext config env m) => Fixity -> Doc ann -> m (Doc ann)
 encloseM fixity doc =
-    view (prettyConfig . renderContext) <&> \context ->
-        encloseIn parens context fixity doc
+  view (prettyConfig . renderContext) <&> \context ->
+    encloseIn parens context fixity doc
 
 -- | The type of a general @config@-based pretty-printer.
-type AnyToDoc config ann = forall a. PrettyBy config a => a -> Doc ann
+type AnyToDoc config ann = forall a. (PrettyBy config a) => a -> Doc ann
 
 -- | Instantiate a supplied continuation with a precedence-aware pretty-printer.
 withPrettyIn
-    :: MonadPrettyContext config env m
-    => ((forall a. PrettyBy config a => Direction -> Fixity -> a -> Doc ann) -> m r) -> m r
+  :: (MonadPrettyContext config env m)
+  => ((forall a. (PrettyBy config a) => Direction -> Fixity -> a -> Doc ann) -> m r) -> m r
 withPrettyIn cont = do
-    config <- view prettyConfig
-    cont $ \dir fixity -> prettyBy $ config & renderContext .~ RenderContext dir fixity
+  config <- view prettyConfig
+  cont $ \dir fixity -> prettyBy $ config & renderContext .~ RenderContext dir fixity
 
--- | Instantiate a supplied continuation with a pretty-printer specialized to supplied
--- 'Fixity' and 'Direction'.
+{-| Instantiate a supplied continuation with a pretty-printer specialized to supplied
+'Fixity' and 'Direction'.
+-}
 withPrettyAt
-    :: MonadPrettyContext config env m
-    => Direction -> Fixity -> (AnyToDoc config ann -> m r) -> m r
+  :: (MonadPrettyContext config env m)
```

Comment:
Is there any option to make it not add the redundant parens?

Reply:
Yes:
```yaml
# Whether to put parentheses around a single constraint (choices: auto, always, or never)
single-constraint-parens: always
```